src/coreclr/jit/lower.cpp

```diff
@@ -7802,6 +7802,312 @@ void Lowering::ContainCheckBitCast(GenTree* node)
+{
+    // Don't merge volatile stores.
+    if (ind->IsVolatile())
+    {
+        return false;
+    }
+
+    // Data has to be INT_CNS, can be also VEC_CNS in future.
+    if (!ind->Data()->IsCnsIntOrI())
+    {
+        return false;
+    }
+
+    // We will need to walk previous trees (up to 4 nodes in a store)
+    // We already know that we have value and addr nodes, LEA might add more.
+    const int MaxTrees        = 4;
+    GenTree*  trees[MaxTrees] = {ind->Data(), ind->Addr(), nullptr, nullptr};
+    int       treesCount      = 2;
+
+    data->targetType = ind->TypeGet();
+    data->value      = ind->Data();
+    if (ind->Addr()->OperIs(GT_LEA))
+    {
+        GenTree* base  = ind->Addr()->AsAddrMode()->Base();
+        GenTree* index = ind->Addr()->AsAddrMode()->Index();
+        if ((base == nullptr) || !base->OperIs(GT_LCL_VAR) || comp->lvaVarAddrExposed(base->AsLclVar()->GetLclNum()))
+        {
+            // Base must be a local. It's possible for it to be nullptr when index is not null,
+            // but let's ignore such cases.
+            return false;
+        }
+
+        if ((index != nullptr) &&
+            (!index->OperIs(GT_LCL_VAR) || comp->lvaVarAddrExposed(index->AsLclVar()->GetLclNum())))
+        {
+            // Index should be either nullptr or a local.
+            return false;
+        }
+
+        data->baseAddr = base == nullptr ? nullptr : base;
+        data->index    = index == nullptr ? nullptr : index;
+        data->scale    = ind->Addr()->AsAddrMode()->GetScale();
+        data->offset   = ind->Addr()->AsAddrMode()->Offset();
+
+        // Add index and base to the list of trees.
+        trees[treesCount++] = base;
+        if (index != nullptr)
+        {
+            trees[treesCount++] = index;
+        }
+    }
+    else if (ind->Addr()->OperIs(GT_LCL_VAR) && !comp->lvaVarAddrExposed(ind->Addr()->AsLclVar()->GetLclNum()))
+    {
+        // Address is just a local, no offset, scale is 1
+        data->baseAddr = ind->Addr();
+        data->index    = nullptr;
+        data->scale    = 1;
+        data->offset   = 0;
+    }
+    else
+    {
+        // Address is not LEA or local.
+        return false;
+    }
+
+    // We expect all the trees to be found in gtPrev (we don't care about the order, just make sure we don't have
+    // unexpected trees in-between)
+    GenTree* prev = ind->gtPrev;
+    for (int i = 0; i < treesCount; i++)
+    {
+        bool found = false;
+
+        // We don't need any O(1) lookups here, the array is too small
+        for (int j = 0; j < MaxTrees; j++)
+        {
+            if (trees[j] == prev)
+            {
+                trees[j] = nullptr;
+                found    = true;
+                break;
+            }
+        }
+
+        if (!found)
+        {
+            // Unexpected node in-between.
+            return false;
+        }
+        prev = prev->gtPrev;
+    }
+
+    *prevTree = prev;
```

Comment:
I think most of this logic could be simplified by using `BlockRange::GetTreeRange` (if the range is not closed return false). The logic here is sort of a manual implementation of what that function does generally.

I mainly ask because I was thinking about what happens when the first node of the tree is the first node in the block. Doesn't `prev` become `nullptr` here?

Reply:
Ah, nice! I wasn't aware of it. Addressed! It saved a ton.